File: 988. Smallest String Starting From Leaf.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func smallestFromLeaf(root *TreeNode) string {
	answer := string('z' + 1)
	var dfs func(*TreeNode, string)
	dfs = func(node *TreeNode, result string) {
		if node == nil {
			return
		}
		result = string('a'+node.Val) + result
		if node.Left == nil && node.Right == nil {
			if result < answer {
				answer = result
			}
		}
		dfs(node.Left, result)
		dfs(node.Right, result)
		result = result[1:]
	}
	dfs(root, "")
	return answer
}

File: 988. Smallest String Starting From Leaf.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
        answer = str(chr(ord('z') + 1))
        def dfs(node, result):
            nonlocal answer
            if not node:
                return
            result = chr(node.val + ord('a')) + result
            if not node.left and not node.right:
                answer = min(answer, result)
            dfs(node.left, result)
            dfs(node.right, result)
            result = result[1:]
        dfs(root, '')
        return answer